### PR TITLE
Allow drop materialized view without storage table

### PR DESCRIPTION
**Pull request by zhaoyim**
allowing user to drop the materialized view instead of
failing after storage table externally dropped
and logging the missing storage table

## Description

> Is this change a fix, improvement, new feature, refactoring, or other?
fix
> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)
iceberg connector
> How would you describe this change to a non-technical end user or system administrator?
allowing user to drop the materialized view instead of failing after storage table externally dropped and logging the missing storage table
## Related issues, pull requests, and links
fixes [12503](https://github.com/trinodb/trino/issues/12503)

## Documentation

(X) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Allow drop materialized view without storage table. ({issue}`12503`)
```

### CODE REVIEW

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractTrinoCatalog.java [hunk 1]
```diff
@@ -81,6 +81,7 @@
     protected static final String PRESTO_VERSION_NAME = HiveMetadata.PRESTO_VERSION_NAME;
     protected static final String PRESTO_QUERY_ID_NAME = HiveMetadata.PRESTO_QUERY_ID_NAME;
     protected static final String PRESTO_VIEW_EXPANDED_TEXT_MARKER = HiveMetadata.PRESTO_VIEW_EXPANDED_TEXT_MARKER;
+    protected static final String STORAGE_TABLE_PREFIX = "st_";
```
Comment (raunaqmorarka) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractTrinoCatalog.java:84]: This change should go into a separate commit

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java [hunk 2]
```diff
@@ -179,6 +179,10 @@ public static Table getIcebergTableWithMetadata(
     public static Map<String, Object> getIcebergTableProperties(Table icebergTable)
     {
         ImmutableMap.Builder<String, Object> properties = ImmutableMap.builder();
+        if (icebergTable == null) {
```
Comment (raunaqmorarka) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java:182]: `getMaterializedViewDefinition` should make the input argument of `icebergTable` an optional and avoid calling this method when it is empty

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergMaterializedViewTest.java [hunk 3]
```diff
@@ -505,6 +505,16 @@ public void testDropMaterializedViewCannotDropTable()
...
     }
 
+    @Test
+    public void testDropMaterializedViewWithoutStorageTable()
+    {
+        assertUpdate("CREATE MATERIALIZED VIEW test_drop_materialized_view_without_storage_table AS SELECT * FROM base_table1");
+        String catalogName = getSession().getCatalog().orElseThrow();
+        SchemaTableName storageTable = getStorageTable(catalogName, "test_drop_materialized_view_without_storage_table");
+        assertUpdate(format("DROP TABLE %s", storageTable.getTableName()));
+        assertUpdate("DROP MATERIALIZED VIEW test_drop_materialized_view_without_storage_table");
```
Comment (raunaqmorarka) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergMaterializedViewTest.java:515]: Please add a few more queries for other operations on MVs like REFRESH, DESCRIBE, listing of MVs through information_schema etc.

**Review by raunaqmorarka — COMMENTED** [review]
It seems that the actual dropMaterializedView code path is already ignoring failure to delete storage table. The root cause here is that the `isMaterializedView` call in `DropMaterializedViewTask` fails because it calls getMV internally which expects the storage table to exist. Same failure would be encountered in listing of MVs or other operations which use isMV or getMV as well.
However, I'm not sure if modifying the getMV code path to return an MV definition with reference to a deleted storage table and no MV properties is safe.
One idea was to catch table-not-found in `MetadataManager#isMaterializedView` but that can be due to MV definition not found as well and there isn't separate error code for that. Another idea is to let connectors Implement isMaterializedView in ConnectorMetadata and decide what to do about the exception there.
@findepi @alexjo2144 thoughts ?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/TrinoHiveCatalog.java [hunk 4]
```diff
@@ -536,6 +536,17 @@ protected Optional<ConnectorMaterializedViewDefinition> doGetMaterializedView(Co
             // - org.apache.iceberg.exceptions.NotFoundException when accessing manifest file
             // - other failures when reading storage table's metadata files
             // Retry, as we're catching broadly.
+            if ((e instanceof io.trino.spi.connector.TableNotFoundException) &&
+                    ((TableNotFoundException) e).getTableName().getTableName().startsWith(STORAGE_TABLE_PREFIX)) {
```
[the remark this answers is not in the document]
Comment (zhaoyim) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/TrinoHiveCatalog.java:540]: yes, you are correct, _TableNotFoundException_ is enough.

**Comment by zhaoyim** [comment]
> It seems that the actual dropMaterializedView code path is already ignoring failure to delete storage table. The root cause here is that the `isMaterializedView` call in `DropMaterializedViewTask` fails because it calls getMV internally which expects the storage table to exist. Same failure would be encountered in listing of MVs or other operations which use isMV or getMV as well. However, I'm not sure if modifying the getMV code path to return an MV definition with reference to a deleted storage table and no MV properties is safe. 

@raunaqmorarka  just update my though here, I did NOT change code in the _MetadataManager_ because its the public class for all the connector, diff connector maybe have the diff implements(such as, seems not all the connector need the storage table to implement), I keep the changes in connector. If a MV without a storage table, I think it's should be an invalid MV, in my opinion, no properties is safe.   

>One idea was to catch table-not-found in `MetadataManager#isMaterializedView` but that can be due to MV definition not found as well and there isn't separate error code for that. Another idea is to let connectors Implement isMaterializedView in ConnectorMetadata and decide what to do about the exception there. @findepi @alexjo2144 thoughts ?

There 2 ideas, I prefer the second, let connectors Implement isMaterializedView. Thanks!

**Comment by covalesj** [comment]
Bump on this -- it looks like it was almost done, and this is an ongoing issue where you cannot drop materialized views when the underlying storage table has been externally dropped.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/TrinoHiveCatalog.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/hms/TrinoHiveCatalog.java:540]: if so, can the changes here be reverted?

**Comment by mosabua** [comment]
:wave: @zhaoyim @raunaqmorarka @findepi  - this PR has become inactive. If you're still interested in working on it, please let us know.

We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.

**Comment by zhaoyim** [comment]
@mosabua Thanks for reminder! You can unassign this issue from me. I am busy on other internal project, Sorry!
